File: games/tank_demolition/js/tank.js
/**
 * Tank class for Tank Demolition Game
 * Handles tank rendering, movement, and rotation in top-down Atari style
 */
class Tank {
    /**
     * Initialize tank at given position
     * @param {number} x - Starting x position (pixels)
     * @param {number} y - Starting y position (pixels)
     */
    constructor(x, y) {
        this.x = x;
        this.y = y;
        this.rotation = 0; // Tank body angle in degrees (0 = up, 90 = right, etc.)
        this.turretRotation = 0; // Turret angle (independent from body)
        this.speed = 2; // Pixels per frame when moving

        // Visual properties - Atari style pixelated tank
        this.width = 48;
        this.height = 48;
        this.tankGreen = '#60a050'; // Bright green like Atari
        this.tankDark = '#409030'; // Dark green for details
        this.trackColor = '#306020'; // Even darker for treads
        this.barrelLength = 30; // Length of turret barrel

        // Movement state
        this.movingForward = false;
        this.movingBackward = false;
        this.rotatingLeft = false;
        this.rotatingRight = false;
        this.rotatingTurretLeft = false;
        this.rotatingTurretRight = false;
    }

    /**
     * Move tank in the direction it's facing
     * @param {boolean} forward - If true, move forward; if false, move backward
     */
    move(forward = true) {
        const direction = forward ? 1 : -1;
        // Convert rotation to radians for math
        const rad = (this.rotation * Math.PI) / 180;

        // Calculate movement based on rotation
        // For top-down view: 0° = up, 90° = right, 180° = down, 270° = left
        this.x += direction * this.speed * Math.sin(rad);
        this.y += direction * this.speed * -Math.cos(rad);
    }

    /**
     * Rotate tank
     * @param {boolean} clockwise - If true, rotate right; if false, rotate left
     */
    rotate(clockwise = true) {
        const rotationSpeed = 3; // Degrees per frame
        if (clockwise) {
            this.rotation += rotationSpeed;
        } else {
            this.rotation -= rotationSpeed;
        }

        // Keep rotation in 0-360 range
        this.rotation = ((this.rotation % 360) + 360) % 360;
    }

    /**
     * Rotate turret independently from tank body
     * @param {boolean} clockwise - If true, rotate right; if false, rotate left
     */
    rotateTurret(clockwise = true) {
        const rotationSpeed = 3; // Degrees per frame
        if (clockwise) {
            this.turretRotation += rotationSpeed;
        } else {
            this.turretRotation -= rotationSpeed;
        }

        // Keep rotation in 0-360 range
        this.turretRotation = ((this.turretRotation % 360) + 360) % 360;
    }

    /**
     * Get the barrel end position in world coordinates
     * @returns {{x: number, y: number}} The barrel end position
     */
    getBarrelEnd() {
        const rad = (this.turretRotation * Math.PI) / 180;
        return {
            x: this.x + this.barrelLength * Math.sin(rad),
            y: this.y + this.barrelLength * -Math.cos(rad)
        };
    }

    /**
     * Prevent tank from moving off screen
     * @param {number} screenWidth - Width of game canvas
     * @param {number} screenHeight - Height of game canvas
     */
    keepInBounds(screenWidth, screenHeight) {
        const margin = 30; // Keep tank this many pixels from edge
        this.x = Math.max(margin, Math.min(screenWidth - margin, this.x));
        this.y = Math.max(margin, Math.min(screenHeight - margin, this.y));
    }

    /**
     * Draw tank in classic Atari top-down style with pixelated look
     * @param {CanvasRenderingContext2D} ctx - Canvas rendering context
     */
    draw(ctx) {
        ctx.save();

        // Translate to tank position and rotate BODY only
        ctx.translate(this.x, this.y);
        ctx.rotate((this.rotation * Math.PI) / 180);

        // Disable image smoothing for pixelated look
        ctx.imageSmoothingEnabled = false;

        // Draw tank treads/tracks (left and right sides) - scaled 1.5x
        ctx.fillStyle = this.trackColor;
        // Left track
        ctx.fillRect(-21, -24, 9, 48);
        // Right track
        ctx.fillRect(12, -24, 9, 48);

        // Draw tank body (main hull) - scaled 1.5x
        ctx.fillStyle = this.tankGreen;
        ctx.fillRect(-15, -18, 30, 36);

        // Draw darker detail on body (hatch/panel) - scaled 1.5x
        ctx.fillStyle = this.tankDark;
        ctx.fillRect(-9, -12, 18, 24);

        // Add pixelated highlights for 3D effect - scaled 1.5x
        ctx.fillStyle = '#70b060';
        ctx.fillRect(-12, -15, 3, 3); // Left highlight
        ctx.fillRect(-12, -6, 3, 3);
        ctx.fillRect(-12, 3, 3, 3);

        // Add shadows for depth - scaled 1.5x
        ctx.fillStyle = this.trackColor;
        ctx.fillRect(9, 12, 3, 3); // Bottom right shadow
        ctx.fillRect(9, 3, 3, 3);
        ctx.fillRect(9, -6, 3, 3);

        ctx.restore();

        // Draw turret separately (rotates independently)
        ctx.save();
        ctx.translate(this.x, this.y);
        ctx.rotate((this.turretRotation * Math.PI) / 180);

        // Draw tank turret barrel (pointing in turret direction)
        ctx.fillStyle = this.tankDark;
        ctx.fillRect(-3, -30, 6, 18);

        // Draw turret base
        ctx.fillStyle = this.tankGreen;
        ctx.fillRect(-9, -9, 18, 18);

        // Draw center dot to show turret direction
        ctx.fillStyle = this.tankDark;
        ctx.fillRect(-2, -2, 4, 4);

        ctx.restore();
    }

    /**
     * Update tank state based on current inputs
     * Called once per frame
     */
    update() {
        if (this.movingForward) {
            this.move(true);
        }
        if (this.movingBackward) {
            this.move(false);
        }
        if (this.rotatingLeft) {
            this.rotate(false);
        }
        if (this.rotatingRight) {
            this.rotate(true);
        }
        if (this.rotatingTurretLeft) {
            this.rotateTurret(false);
        }
        if (this.rotatingTurretRight) {
            this.rotateTurret(true);
        }
    }
}

File: games/tank_demolition/tank.py
"""
Tank class for Tank Demolition Game
Handles tank rendering, movement, and rotation in isometric view
"""
import pygame
import math


class Tank:
    """
    Represents the player's tank with movement and rotation capabilities.
    Rendered in isometric perspective with simple shading.
    """

    def __init__(self, x, y):
        """
        Initialize tank at given position.

        Args:
            x: Starting x position (pixels)
            y: Starting y position (pixels)
        """
        self.x = x
        self.y = y
        self.rotation = 0  # Angle in degrees (0 = right, 90 = down, etc.)
        self.speed = 3  # Pixels per frame when moving

        # Visual properties
        self.width = 60
        self.height = 40
        self.body_color_top = (80, 120, 60)  # Lighter military green
        self.body_color_side = (60, 90, 45)  # Darker military green
        self.shadow_color = (0, 0, 0, 100)  # Semi-transparent black

        # Movement state
        self.moving_forward = False
        self.moving_backward = False
        self.rotating_left = False
        self.rotating_right = False

    def move(self, forward=True):
        """
        Move tank in the direction it's facing.

        Args:
            forward: If True, move forward; if False, move backward
        """
        direction = 1 if forward else -1
        # Convert rotation to radians for math
        rad = math.radians(self.rotation)

        # Calculate movement based on rotation
        self.x += direction * self.speed * math.cos(rad)
        self.y += direction * self.speed * math.sin(rad)

    def rotate(self, clockwise=True):
        """
        Rotate tank.

        Args:
            clockwise: If True, rotate right; if False, rotate left
        """
        rotation_speed = 3  # Degrees per frame
        if clockwise:
            self.rotation += rotation_speed
        else:
            self.rotation -= rotation_speed

        # Keep rotation in 0-360 range
        self.rotation = self.rotation % 360

    def keep_in_bounds(self, screen_width, screen_height):
        """
        Prevent tank from moving off screen.

        Args:
            screen_width: Width of game window
            screen_height: Height of game window
        """
        margin = 30  # Keep tank this many pixels from edge
        self.x = max(margin, min(screen_width - margin, self.x))
        self.y = max(margin, min(screen_height - margin, self.y))

    def draw_shadow(self, surface):
        """
        Draw shadow under tank.

        Args:
            surface: Pygame surface to draw on
        """
        # Create semi-transparent surface for shadow
        shadow_surface = pygame.Surface((self.width + 10, self.height // 2), pygame.SRCALPHA)

        # Draw oval shadow
        pygame.draw.ellipse(
            shadow_surface,
            self.shadow_color,
            (0, 0, self.width + 10, self.height // 2)
        )

        # Position shadow slightly offset from tank
        shadow_x = self.x - (self.width + 10) // 2 + 5
        shadow_y = self.y - (self.height // 2) // 2 + 10

        surface.blit(shadow_surface, (shadow_x, shadow_y))

    def draw(self, surface):
        """
        Draw tank in isometric view with shading.

        Args:
            surface: Pygame surface to draw on
        """
        # Draw shadow first (underneath tank)
        self.draw_shadow(surface)

        # Create a surface for the tank body
        tank_surface = pygame.Surface((self.width, self.height), pygame.SRCALPHA)

        # Draw isometric tank body (two rectangles for depth)
        # Top face (lighter)
        top_points = [
            (self.width // 4, 0),
            (3 * self.width // 4, 0),
            (self.width, self.height // 3),
            (0, self.height // 3)
        ]
        pygame.draw.polygon(tank_surface, self.body_color_top, top_points)

        # Side face (darker)
        side_points = [
            (0, self.height // 3),
            (self.width, self.height // 3),
            (self.width, self.height),
            (0, self.height)
        ]
        pygame.draw.polygon(tank_surface, self.body_color_side, side_points)

        # Add outline for definition
        pygame.draw.polygon(tank_surface, (40, 60, 30), top_points, 2)
        pygame.draw.polygon(tank_surface, (40, 60, 30), side_points, 2)

        # Rotate the tank surface
        rotated_tank = pygame.transform.rotate(tank_surface, -self.rotation)

        # Get rectangle for positioning (center on tank position)
        tank_rect = rotated_tank.get_rect(center=(self.x, self.y))

        # Draw rotated tank
        surface.blit(rotated_tank, tank_rect)

    def update(self):
        """
        Update tank state based on current inputs.
        Called once per frame.
        """
        if self.moving_forward:
            self.move(forward=True)
        if self.moving_backward:
            self.move(forward=False)
        if self.rotating_left:
            self.rotate(clockwise=False)
        if self.rotating_right:
            self.rotate(clockwise=True)
